documentation/projects/proposals/nuxt_3_migration/20230608-implementation_plan_nuxt_3_migration.md

```diff
@@ -0,0 +1,338 @@
+
+### Nuxt 3 / Vue 3 features in Openverse: discussion
+
+#### Autoimports
+
+Nuxt 3 auto imports composables and components. Unlike the usual global imports,
+these imports are handled in such a way that only the imports that are actually
+used are left in the production files. The Nuxt team have optimized the way they
+work both in the working app and in the IDEs, so I think we _should_ use the
+auto import feature and not go against the framework and disable it. In the
+current setup, our test suite throws warnings when a child component is not
+imported. We would need to make sure that this does not happen with Nuxt 3.
+
+#### `script setup`
+
+The new way of writing components in Vue 3 is using `script setup` which reduces
+the number of code lines by automatically exporting the variables for use in the
+components. While we can eventually move to using `script setup` throughout the
+app, this conversion is out of scope for this project. We can, however, convert
+some components if they need extensive code changes anyway, or if the behavior
+within them changes significantly and the `script setup` simplifies the
+component. The decision whether to convert components to `script setup` would be
+at the discretion of the PR implementer, but e should not refactor for
+refactoring's sake alone, after all.
+
+#### Testing
+
+Vue 3 recommends to use Vitest for better performance and integration with
+`vite` (which will replace `webpack` in our setup). I have not had an
+opportunity to test it out, but I think it would make sense to convert our unit
+tests to `Vitest`. The decision on whether to keep using the current
+dependencies or convert to `Vitest` should be based on the ease of conversion.
+[Type tests is also interesting and may be useful in some places.](https://vitest.dev/guide/testing-types.html)
+We should add
+[`eslint-plugin-vitest`](https://github.com/veritem/eslint-plugin-vitest#credits)
+which basically a port of eslint-plugin-jest with necessary modifications.
+
+## Expected Outcomes
+
+<!-- List any succinct expected products from this implementation plan. -->
+
+Openverse frontend uses Nuxt 3, and all current features (including i18n,
+sitemaps, SEO) work.
+
+## Outlined Steps
+
+<!-- Describe the implementation step necessary for completion. -->
+
+This section describes the known steps that need to be taken to migrate to
+Nuxt 3. However, it is possible that there will be some unknown steps that will
+be discovered during the implementation.
+
+### Changes to main
+
+- Deprecation of CJS modules (the ones that use `require` and `module.exports`):
+
+  - [ ] `case` package that we use for `kebab-case`, `title` case and
+        `camelCase` function is CJS. We can either find replacements that use
+        `ESM` modules, vendor in these functions, or use a workaround that
+        `vite` suggests (instead of `const { kebab } = require "case"` use
+        `import casePkg; const { kebab } = casePkg;`)
```

Comment:
I think we could just vendor this and recommend that, rather than listing multiple potential options here.